### PR TITLE
Kevin/listening duration

### PR DESCRIPTION
**Pull request by kevingreer**
This is a change to listening history so that each entry contains (1) the percentage of the episode that has been listened to as a float (could be higher than 1 if listened to more than once) and (2) the percentage of the episode the user is currently at as a float (i.e. they stopped listening 76% through and should pick up there).

Also this includes a new field in the episode model: real_duration. The regular duration field could be in an inconsistent, so client can send us a float value indicating the true length of the episode in seconds.

### CODE REVIEW

File: src/app/pcasts/controllers/create_delete_listening_history_controller.py [hunk 1]
```diff
@@ -6,17 +6,13 @@ def get_path(self):
     return '/history/listening/<episode_id>/'
```
Comment (ChenJesse) [inline src/app/pcasts/controllers/create_delete_listening_history_controller.py]: rename file?

File: src/app/pcasts/dao/listening_histories_dao.py [hunk 2]
```diff
@@ -1,25 +1,38 @@
...
+                                  ListeningHistory.user_id == user.id).all()
+  listening_histories_by_episode = \
+    {lh.episode_id: lh for lh in listening_histories}
+  result = []
+  for e in episodes:
+    try:
+      current_progress = episode_update_info_map[e.id]['current_progress']
+      percentage_listened = episode_update_info_map[e.id]['percentage_listened']
+    except KeyError:
+      raise Exception('Invalid episode_id {} provided'.format(e.id))
```
Comment (ChenJesse) [inline src/app/pcasts/dao/listening_histories_dao.py]: Should an exception be raised here? Or maybe it should try to create/update all the episodes that have a valid episode_id? I'm not sure cause idk the case where the episode_id would ever be invalid

File: src/app/pcasts/dao/listening_histories_dao.py [hunk 3]
```diff
@@ -1,25 +1,38 @@
...
-      optional_listening_history.update_listening_time()
-      db_utils.db_session_commit()
-      return optional_listening_history
-
-    listening_history = ListeningHistory(episode_id=episode_id, user_id=user.id)
-    return db_utils.commit_model(listening_history)
-  else:
-    raise Exception('Invalid episode_id provided')
+def create_or_update_listening_histories(episode_update_info_map, user):
+  episode_ids = episode_update_info_map.keys()
```
Comment (ChenJesse) [inline src/app/pcasts/dao/listening_histories_dao.py:5]: Not sure the rate at which we will have `real_duration` in the `episode_update_info_map` but if it is not very often, it might make more sense to get the episodes objects on demand, instead of doing a batch. Otherwise logic is really solid

**Review by ChenJesse — APPROVED** [review]
Logic looks solid to me, the only thing is the name `real_duration` is a little confusing, maybe something like `duration_seconds` would be more readable. Also, I'm not sure if it's better to throw an exception for everything if we find a single invalid episode_id, or to just try to complete all the valid ones. Ultimately up to you

File: src/app/pcasts/dao/listening_histories_dao.py (hunk 2 above)
Comment (kevingreer) [reply src/app/pcasts/dao/listening_histories_dao.py]: Yeah idk what best practices for batches are, but I feel like if some part of the batch is invalid it might be better to not create anything because episode_id would only be invalid if there was a bug on client

File: src/app/pcasts/controllers/listening_history_controller.py [hunk 4]
```diff
@@ -0,0 +1,31 @@
...
+      max_hs = int(request.args['max'])
+      histories = \
+        listening_histories_dao.get_listening_history(user, max_hs, offset)
+
+    elif request.method == 'POST':
+      user = kwargs.get('user')
+      episode_info = {long(k): v for k, v in json.loads(request.data).items()}
+      histories = listening_histories_dao \
+        .create_or_update_listening_histories(episode_info, user)
+      app.logger.info(
```
Comment (ifilonenko) [inline src/app/pcasts/controllers/listening_history_controller.py:26]: Should the logger be pushing user_id information? Isn't that dangerous from a security standpoint?

File: src/app/pcasts/dao/listening_histories_dao.py [hunk 5]
```diff
@@ -1,25 +1,38 @@
...
+  episodes = episodes_dao.get_episodes(episode_ids, user.id)
```
Comment (ifilonenko) [inline src/app/pcasts/dao/listening_histories_dao.py]: why would you need to pass in `user.id` into `episodes`?

File: src/app/pcasts/dao/listening_histories_dao.py [hunk 6]
```diff
@@ -1,25 +1,38 @@
...
+  for e in episodes:
+    try:
+      current_progress = episode_update_info_map[e.id]['current_progress']
+      percentage_listened = episode_update_info_map[e.id]['percentage_listened']
+    except KeyError:
+      raise Exception('Invalid episode_id {} provided'.format(e.id))
+    real_duration = \
+      None if 'real_duration' not in episode_update_info_map[e.id] \
+      else episode_update_info_map[e.id]['real_duration']
+    if e.id in listening_histories_by_episode:
```
Comment (ifilonenko) [inline src/app/pcasts/dao/listening_histories_dao.py]: `for (e,episode) in listening_histories_by_episode.items():`

**Review by ifilonenko — CHANGES_REQUESTED** [review]
Make changes requested above.

File: src/app/pcasts/dao/listening_histories_dao.py (hunk 5 above)
Comment (ChenJesse) [reply src/app/pcasts/dao/listening_histories_dao.py]: client wants us to give them a fields like `is_bookmarked` and `is_recommended` from the perpsective of whoever is making the request for convenience purposes

Comment (ifilonenko) [reply src/app/pcasts/dao/listening_histories_dao.py]: Is this necessary for the listening_histories?

File: src/app/pcasts/controllers/listening_history_controller.py (hunk 4 above)
Comment (kevingreer) [reply src/app/pcasts/controllers/listening_history_controller.py:26]: maybe not, but every endpoint that creates or deletes something does this, so that could be changed in a future PR

Comment (ChenJesse) [reply src/app/pcasts/controllers/listening_history_controller.py:26]: the sessions should prevent any malicious api calls, and I'm not too knowledgeable about how secure logs on a server are, but i guess it doesnt help. I just thought it would be convenient for debugging purposes to know the id (if we wanna make some db requests), but we can definitely do without it if its dangerous

File: src/app/pcasts/dao/listening_histories_dao.py (hunk 5 above)
Comment (ChenJesse) [reply src/app/pcasts/dao/listening_histories_dao.py]: dont think so, maybe we can make that parameter optional and not attach those fields if nothing is passed in? it would make it slightly more efficient, but we should prob do that in another pr

File: src/app/pcasts/controllers/listening_history_controller.py (hunk 4 above)
Comment (ifilonenko) [reply src/app/pcasts/controllers/listening_history_controller.py:26]: Production PRs should not include doing logs of DB specific ids.

Comment (kevingreer) [reply src/app/pcasts/controllers/listening_history_controller.py:26]: so what do you think we should log, if anything?

File: src/app/pcasts/dao/listening_histories_dao.py (hunk 5 above)
Comment (kevingreer) [reply src/app/pcasts/dao/listening_histories_dao.py]: since we currently return the list of affected `ListeningHistory` objects, I think they do need the bookmarked and recommended field in case client wants to do anything with the resulting episodes. If they don't use the result, then we could return nothing and not do the computation to tack that info onto the episodes.

File: src/app/pcasts/controllers/listening_history_controller.py (hunk 4 above)
Comment (ChenJesse) [reply src/app/pcasts/controllers/listening_history_controller.py:26]: also, I'm curious why logging id's isn't best practice? I don't understand how the logs would be accessed by malicious users in the first place, and even if they could, what could they do with it, since identity is confirmed through a session?

**Comment by kevingreer** [comment]
I just changed the structure of `real_duration`. The `real_duration` field is gone, instead client sends us the real duration in the exact string format they will use in the future. We overwrite the `duration` field on the episode, and set a new boolean field `real_duration_written` that will be checked so we don't waste energy overwriting the duration more than once. Client will check that field and determine whether or not they should send us the real_duration. 

The benefit of this method is that the additional boolean field takes less space than the double field, and now we only query the episodes table for the episodes that need to be updated

File: src/app/pcasts/controllers/listening_history_controller.py (hunk 4 above)
Comment (ifilonenko) [reply src/app/pcasts/controllers/listening_history_controller.py:26]: Because malicious users who have access to server logs should not have awareness of DB specific information. For example: printing in logs credit card information after successful submission. You wouldnt log that, as such you shouldn’t print in production similarly. For debugging purposes, logs are useful but shouldn’t be leveraged in production instances for the sake of security guarantees of our service

Comment (ChenJesse) [reply src/app/pcasts/controllers/listening_history_controller.py:26]: Yeah credit card info definitely wouldnt make much sense b/c with it alone, you can do some malicious stuff, but I guess outside of the context of our app, user id doesnt mean as much, I cant think of what they could do with it alone. And I thought the entire point of logging was primarily for production instances, so that if there is some bug we can trace down the calls that lead up to it, given user who reports it and timestamp. Not sure how useful they would be just in such a controlled test environment, if that makes sense. At least from what I saw from my (limited) experience in industry so far, we used logs heavily in production instances to track down bugs reported by users

After doing some research, I'm pretty sure my experience wasn't an exception; from this nextdoor engineering blog (https://engblog.nextdoor.com/what-to-log-in-production-environments-cc9ad82bdce9):

"Logging is a crucial tool for any application, because it won’t always be easy to reproduce issues in development where you can use an interactive debugger. Being able to control logging at runtime lets us spend less time trying to set up the conditions to reproduce issues in dev and more time writing code."

Also, if malicious users have access to server logs, doesnt that mean they have access to the entire server? in that case, we would have bigger things to worry about, like all the secrets and env variables right? I think we are supposed to preven tthat from happening in the first place

Comment (ifilonenko) [reply src/app/pcasts/controllers/listening_history_controller.py:26]: What I am trying to say is that logging of DB specific stuff. like DB internal IDs, env variables, etc. is security flaw. Things that people can reverse engineer with malicious intent. If you feel as though the user id serves no risk, than it is okay to include in the logs. The problem is when you start logging information that sys admins shouldn't necessarily have access to in the case that a sys admin becomes malicious

Comment (ChenJesse) [reply src/app/pcasts/controllers/listening_history_controller.py:26]: Ok that's a fair point, at the current state, user id doesn't maintain any risk with the endpoints we have currently, but I guess there is no guarantee for that in the future, and it is redundant with the username being logged as well. I'll make another PR to go through and remove all the user_ids, and I also plan on logging in json format instead of raw string, should be easier to query for stuff that way

File: src/app/pcasts/dao/listening_histories_dao.py [hunk 7]
```diff
@@ -3,34 +3,42 @@
...
+  # Update episode durations if needed
+  episode_ids_to_query = []
+  for episode_id, values in episode_update_info_map.iteritems():
+    if 'real_duration' in values:
+      episode_ids_to_query.append(episode_id)
+  episodes = episodes_dao.get_episodes(episode_ids_to_query, user.id)
+  for episode in episodes:
+    if not episode.real_duration_written:
+      episode.duration = episode_update_info_map[episode.id]['real_duration']
+      episode.real_duration_written = True
```
Comment (ifilonenko) [inline src/app/pcasts/dao/listening_histories_dao.py:39]: Why is this necessary? You are sequentially reading through list. How would you go through this list twice?

File: tests/test_listening_history.py [hunk 8]
```diff
@@ -18,18 +19,26 @@ def setUp(self):
   def tearDown(self):
     super(ListeningHistoryTestCase, self).tearDown()
     ListeningHistory.query.delete()
-    for e in self.changed_episodes:
-      e.real_duration = None
+    episodes = \
+      Episode.query.filter(Episode.id.in_(self.changed_episodes.keys())).all()
+    for episode in episodes:
```
Comment (ifilonenko) [inline tests/test_listening_history.py]: If you are only grabbing id, why are you serializing entire episodes? Can you possibly in the query return just the ids?

Comment (ifilonenko) [reply tests/test_listening_history.py]: oh these are tests... awks... :)

File: src/app/pcasts/dao/listening_histories_dao.py [hunk 9]
```diff
@@ -1,25 +1,46 @@
...
+    else:
+      lh = \
+        ListeningHistory(episode_id=episode_id, user_id=user.id,
+                         percentage_listened=percentage_listened,
+                         current_progress=current_progress)
+      db.session.add(lh)
+    result.append(lh)
+  # Update episode durations if needed
+  episode_ids_to_query = []
+  for episode_id, values in episode_update_info_map.iteritems():
```
Comment (ifilonenko) [inline src/app/pcasts/dao/listening_histories_dao.py]: Maybe try re-writing this with zip() and by filtering on keys. Use some functional programming concepts here instead of this savage array work here, checking on values.

**Review by ifilonenko — APPROVED** [review]

File: src/app/pcasts/dao/listening_histories_dao.py (hunk 7 above)
Comment (kevingreer) [reply src/app/pcasts/dao/listening_histories_dao.py:39]: the real_duration_written might have been set to True in previous batch

File: src/app/pcasts/dao/listening_histories_dao.py (hunk 9 above)
Comment (kevingreer) [reply src/app/pcasts/dao/listening_histories_dao.py]: Nice point, this can be compressed to a pretty simple `[k for k, v in episode_update_info_map.items() if 'real_duration' in v]`
